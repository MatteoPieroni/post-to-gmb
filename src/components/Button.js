import React, { Component } from 'react';
import PropTypes from 'prop-types';

const Button = ({addedClass, buttonType = "button", onClick, text}) => {
	return (
		<button className={ 'button ' + addedClass } type={ buttonType } onClick={ onClick }>{text}</button>
	)
}

Button.propTypes = {
	addedClass: PropTypes.string,
	buttonType: PropTypes.string,
	onClick: PropTypes.func,
	text: PropTypes.string.isRequired,
};

export default Button;