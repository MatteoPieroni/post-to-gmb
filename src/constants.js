export const button = {
	notSpecified: "ACTION_TYPE_UNSPECIFIED",
	book: "BOOK",
	order: "ORDER",
	shop: "SHOP",
	learn: "LEARN_MORE",
	signUp: "SIGN_UP",
	getOffer: "GET_OFFER",
	call: "CALL",
}

export const postState = {
	notSpecified: "LOCAL_POST_STATE_UNSPECIFIED",
	rejected: "REJECTED",
	live: "LIVE",
	processing: "PROCESSING",
}

export const topicType = {
	notSpecified: "LOCAL_POST_TOPIC_TYPE_UNSPECIFIED",
	standard: "STANDARD",
	event: "EVENT",
	offer: "OFFER",
	product: "PRODUCT",
}

export const placeStatus = {
	open: "OPEN",
}