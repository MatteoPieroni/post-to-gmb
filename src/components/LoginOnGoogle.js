import React, {Component} from 'react';
import PropTypes from 'prop-types';

import {apiData} from '../apiData';
import Button from './Button';
import {gapiWrapper} from './gapiWrapper';

const gapi = gapiWrapper.init();

class LoginOnGoogle extends Component {
	constructor(props) {
		super(props);

		this.state = {
			loggedIn: null,
		}

		this.initGoogleClient = this.initGoogleClient.bind(this);
		this.updateSigninStatus = this.updateSigninStatus.bind(this);
		this.logIn = this.logIn.bind(this);
		this.logOut = this.logOut.bind(this);
	}

	componentDidMount() { 
		if(gapi)
	    	gapi.load('client:auth2', this.initGoogleClient);
	}

	initGoogleClient () {
		const component = this;

		gapi.client.init({
            apiKey: apiData.apiKey,
            clientId: apiData.clientId,
            scope: apiData.scopes
        }).then(function() {
          // Listen for sign-in state changes.
          gapi.auth2.getAuthInstance().isSignedIn.listen(component.updateSigninStatus);

          // Handle the initial sign-in state.
          component.updateSigninStatus(gapi.auth2.getAuthInstance().isSignedIn.get());
      })
    };

	updateSigninStatus(isSignedIn) {
		if(isSignedIn) {
			this.setState({
				loggedIn: true,
			});
			this.props.clientReady();
		} else {
			this.setState({
				loggedIn: false,
			});
			this.props.clientNotReady();
		}
	}

	logIn() {
		gapi.auth2.getAuthInstance().signIn();
	}

	logOut() {
		gapi.auth2.getAuthInstance().signOut();
	}

	render() {
		const loggedIn = this.state.loggedIn;
		if (loggedIn === false)
			return <Button onClick={this.logIn} text="Log in on Google" />
		else if (loggedIn === true)
			return <Button onClick={this.logOut} text="Log out" />
		else
			return null
	}
}

LoginOnGoogle.propTypes = {
	clientReady: PropTypes.func.isRequired,
	clientNotReady: PropTypes.func.isRequired,
}

export default LoginOnGoogle;