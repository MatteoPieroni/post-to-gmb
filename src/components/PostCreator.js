import React, {Component} from 'react';
import PropTypes from 'prop-types';
import Select from 'react-select';

import {button} from '../constants';
import Upload from './Upload';
import PostText from './PostText';
import Input from './Input';
import Button from './Button';

class PostCreator extends Component {
	constructor(props) {
		super(props);
		this.state = {
			selectOptions: [
				{
					value: button.notSpecified,
					label: 'Choose button (optional)',
				},

				{
					value: button.book,
					label: 'Book',
				},
				{
					value: button.order,
					label: 'Order',
				},
				{
					value: button.shop,
					label: 'Shop',
				},
				{
					value: button.learn,
					label: 'Learn More',
				},
				{
					value: button.signUp,
					label: 'Sign Up',
				},
				{
					value: button.getOffer,
					label: 'Get Offer',
				},
				{
					value: button.call,
					label: 'Call',
				},
			],
			selectedOption: {
					value: button.notSpecified,
					label: 'Choose button (optional)',
				},
		}

		this.onChangeSelect = this.onChangeSelect.bind(this);
	}


	onChangeSelect = (selectedOption) => {
		this.setState({ selectedOption });
	}

	render() {
		const selectedOption = this.state.selectedOption;
		const options = this.state.selectOptions;
		return (
			<div className={"post-creator__container " + this.props.addedClass}>
				<form id="postCreator" onSubmit={this.props.onSubmit}>
					<Upload onChange={this.props.uploadImage} />
					<PostText onChange={this.props.textChanged} />
					<Select
						name="post-button"
						value={selectedOption}
						onChange={this.onChangeSelect}
						options={options}
					/>
					<Input inputType="text" labelText="Button url" name="post-button-url" />
					<Button disabled={false} buttonType="submit" text="Post to your page" />
				</form>
			</div>
		)
	}
};

PostCreator.propTypes = {
	addedClass: PropTypes.string,
	uploadImage: PropTypes.func.isRequired,
	onSubmit: PropTypes.func.isRequired,
	textChanged: PropTypes.func,
}

export default PostCreator;