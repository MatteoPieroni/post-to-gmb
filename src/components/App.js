import React, { Component } from 'react';
import axios from 'axios';
import './App.css';

import Select from 'react-select';

import * as Constants from '../constants';
import {apiData} from '../apiData';
import {gapiWrapper} from './gapiWrapper';
import Button from './Button';
import PostCreator from './PostCreator'
import LoginOnGoogle from './LoginOnGoogle';
import PageSelector from './PageSelector';

const gapi = gapiWrapper.init();

class App extends Component {

  constructor(props) {
    super(props);

    this.state = {
      languageCode: "it-IT",
      topicType: Constants.topicType.standard,
      selectedLocations: [],
    }

    this.getUserAndToken = this.getUserAndToken.bind(this);
    this.getAccounts = this.getAccounts.bind(this);
    this.getLocations = this.getLocations.bind(this);
    this.selectLocations = this.selectLocations.bind(this);
    this.getDataToSend = this.getDataToSend.bind(this);
    this.postToGMB = this.postToGMB.bind(this);
    this.uploadImage = this.uploadImage.bind(this);
    this.textChangeFunction = this.textChangeFunction.bind(this);
  }

  getUserAndToken() {
    return new Promise((resolve, reject) => {
      let user = gapi.auth2.getAuthInstance().currentUser.get();
      let oauthToken = user.getAuthResponse().access_token;
      if(user && oauthToken) {
        this.setState({
          userInstance: {
            user: user,
            oauthToken: oauthToken,
          }
        });
        axios.defaults.headers.common['Authorization'] = `Bearer ${this.state.userInstance.oauthToken}`;
        axios.defaults.headers.post['Content-Type'] = 'application/json';
        resolve();
      } else {
        let error = 'NOT_CONNECTED';
        reject(error);
      }
    })
  }

  getAccounts() {
    let component = this;

    this.getUserAndToken().then(() => {

      let url = apiData.gmb_api_version + '/accounts';

      axios.get(url).then(function(response) {
        console.log(response);
        component.getLocations(response.data.accounts);
      }).catch(function(err) {
        console.log(err);
      });

    }).catch(error => {
      console.log(error);
    })
  }

  getLocations(accounts) {
    let component = this;
    for(let i = 0; i < accounts.length; i++) {
        let url = apiData.gmb_api_version + '/' + accounts[i].name + '/locations';

        axios.get(url).then(function(response) {
          console.log(response);

          let locations = [];
          response.data.locations.map((location) => {
            if(location.openInfo.status === Constants.placeStatus.open && location.locationState.isVerified === true) {
              locations.push({
                value: location.name,
                label: `${location.locationName} (${(location.storeCode) ? location.storeCode : 'no-sc' })`
              });
            };
          });
          component.setState({
            locations: locations,
          });
        }).catch(function(err) {
          console.log(err);
        });

    }
  }

  selectLocations(selectedOption) {
    let selectedLocations = this.state.selectedLocations;

    /*if(e.target.checked) {
      selectedLocations.push(e.target.id);
    } else {
      let index = selectedLocations.indexOf(e.target.id);
      if (index > -1) {
        selectedLocations.splice(index, 1);
      }
    }*/

    this.setState({
      selectedLocations: selectedOption,
    });
  }

  getDataToSend(event) {
    event.preventDefault();

    const formToSubmit = document.getElementById(event.target.id);
    const formElements = formToSubmit.elements;
    const dataToSend = {};
    dataToSend.languageCode = this.state.languageCode;
    dataToSend.callToAction = {};
    dataToSend.topicType = this.state.topicType;

    for(let i = 0; i < formElements.length; i++) {
      switch(formElements[i].name) {
        case 'post-text':
          dataToSend.summary = formElements[i].value;
          break;
        case 'post-button':
          dataToSend.callToAction.actionType = formElements[i].value;
          break;
        case 'post-button-url':
          dataToSend.callToAction.url = formElements[i].value;
          break;
      }
    }

    if(dataToSend.callToAction.actionType === Constants.button.notSpecified)
      delete dataToSend.callToAction;

    this.postToGMB(dataToSend);

  }

  postToGMB(data) {
    const locationsToSend = this.state.selectedLocations;
    locationsToSend.forEach((el) => {
    let url = apiData.gmb_api_version + '/' + el.value + '/localPosts';

    axios.post(url, JSON.stringify(data)).then(function(response) {
        console.log(response);
      }).catch(function(err) {
        console.log(err);
      });
    })
  }

  uploadImage() {
    console.log('uploading');
  }

  textChangeFunction() {
    console.log('text changed');
  }

  render() {
    const locations = [];
    /*const accountPages = (this.state.locations) ? (this.state.locations.map((location) => {
      //return <PageSelector key={location.name} pageName={location.locationName} pageId={location.name} onChange={this.selectLocations} />
      locations.push({
        value: location.name,
        label: location.locationName
      });
    })) : '';*/
    return (
      <div className="App">
        <LoginOnGoogle clientReady={this.getAccounts} clientNotReady={() => console.log('client not ready')} />
        <header className="App-header">
        </header>
        <div className="page-selector__list">
          <Select options={this.state.locations} isMulti={true} isSearchable={true} onChange={this.selectLocations} />
        </div>
        <PostCreator onSubmit={this.getDataToSend} uploadImage={this.uploadImage} textChanged={this.textChangeFunction} />
      </div>
    );
  }
}

export default App;
