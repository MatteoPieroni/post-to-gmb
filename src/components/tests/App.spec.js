import React from 'react';
import { shallow } from 'enzyme';

import Select from 'react-select';

import * as Constants from '../../constants';
import App from '../App';
import LoginOnGoogle from '../LoginOnGoogle';
import PostCreator from '../PostCreator';
import Button from '../Button';

describe('App', () => {
	let wrapper;
	beforeEach(() => {
		wrapper = shallow(<App />);
	});

	it('should render correctly', () => {
		expect(wrapper).toMatchSnapshot();
	});
	it('should render a wrapping <div /> and an internal div', () => {
		expect(wrapper.find('div').length).toEqual(2);
	});
	it('should render a <Select> element', () => {
		expect(wrapper.containsMatchingElement(<Select />)).toEqual(true);
	});
	it('should display a PostCreate element', () => {
		expect(wrapper.containsMatchingElement(<PostCreator />)).toEqual(true);
	});
	it('should display a LoginOnGoogle element', () => {
		expect(wrapper.containsMatchingElement(<LoginOnGoogle />)).toEqual(true);
	});
	it('should have a default state with topicType of standard', () => {
		expect(wrapper.state('topicType')).toEqual(Constants.topicType.standard);
	});
});