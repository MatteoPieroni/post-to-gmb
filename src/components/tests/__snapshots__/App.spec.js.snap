// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should render correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <App />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <LoginOnGoogle
          clientNotReady={[Function]}
          clientReady={[Function]}
        />,
        <header
          className="App-header"
        />,
        <div
          className="page-selector__list"
        >
          <StateManager
            defaultInputValue=""
            defaultMenuIsOpen={false}
            defaultValue={null}
            isMulti={true}
            isSearchable={true}
            onChange={[Function]}
          />
        </div>,
        <PostCreator
          onSubmit={[Function]}
          textChanged={[Function]}
          uploadImage={[Function]}
        />,
      ],
      "className": "App",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "clientNotReady": [Function],
          "clientReady": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "className": "App-header",
        },
        "ref": null,
        "rendered": null,
        "type": "header",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <StateManager
            defaultInputValue=""
            defaultMenuIsOpen={false}
            defaultValue={null}
            isMulti={true}
            isSearchable={true}
            onChange={[Function]}
          />,
          "className": "page-selector__list",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "defaultInputValue": "",
            "defaultMenuIsOpen": false,
            "defaultValue": null,
            "isMulti": true,
            "isSearchable": true,
            "onChange": [Function],
            "options": undefined,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "onSubmit": [Function],
          "textChanged": [Function],
          "uploadImage": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <LoginOnGoogle
            clientNotReady={[Function]}
            clientReady={[Function]}
          />,
          <header
            className="App-header"
          />,
          <div
            className="page-selector__list"
          >
            <StateManager
              defaultInputValue=""
              defaultMenuIsOpen={false}
              defaultValue={null}
              isMulti={true}
              isSearchable={true}
              onChange={[Function]}
            />
          </div>,
          <PostCreator
            onSubmit={[Function]}
            textChanged={[Function]}
            uploadImage={[Function]}
          />,
        ],
        "className": "App",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "clientNotReady": [Function],
            "clientReady": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "App-header",
          },
          "ref": null,
          "rendered": null,
          "type": "header",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <StateManager
              defaultInputValue=""
              defaultMenuIsOpen={false}
              defaultValue={null}
              isMulti={true}
              isSearchable={true}
              onChange={[Function]}
            />,
            "className": "page-selector__list",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "defaultInputValue": "",
              "defaultMenuIsOpen": false,
              "defaultValue": null,
              "isMulti": true,
              "isSearchable": true,
              "onChange": [Function],
              "options": undefined,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "onSubmit": [Function],
            "textChanged": [Function],
            "uploadImage": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
